### ISSUE TITLE
clang gives false positives on overloaded comparison operators in C++

### ISSUE THREAD

**Issue by akrzemi1** [llvm/llvm-project#80538]
clang++ issues a warning when I have an expression like `a >> b > c`:

```
warning: overloaded operator >> has higher precedence than comparison operator [-Woverloaded-shift-op-parentheses]
```

This warning is generated by default, even if I do not put `-Wall`. It is useful for built-in comparison operators, but it generates false positives when I overload my comparison operators for the purpose of doing "expression templates":

https://godbolt.org/z/a4v41qeYz

```
template <Node T, Node U>
Follows<T, U> operator>(T t, U u) { return {t, u}; }

template <Node T, Node U>
FastFollows<T, U> operator>>(T t, U u) { return {t, u}; }
```

I recommend changing the warning so that it does not show up when the return type of the comparison operator is other than `bool`.

This is a practical problem that shows up when using libraries such as Boost.Spirit:

https://godbolt.org/z/EMqGa1zzh

**Comment by shafik** [comment]
The diagnostic is `warn_overloaded_shift_in_comparison` and it is handled in `DiagnoseShiftCompare`. I would have to look at it in more detail but it looks fixable.
